### PR TITLE
Implement Glimpse Draft format (Closes #1007)

### LINKED ISSUE #1007: Feature request: Implement Glimpse Draft
[Specification](http://turnonemagic.com/glimpse-drafting-best-way-draft-cube-2-4-players-title-shorter/) of Glimpse Draft format.

I'll implement it after I (or someone else) will fix #1006

### PR DESCRIPTION
**Pull request by serokellcao**
Fixes #1007

Problem: There is no skill-testing 2-player draft format

Solution:

 1. Implement glimpse draft as a superset of regular draft
    1. human.js
      - Make autopick into a list of picks to make
      - Do pick after pick, removing elements from this list
      - On new cycle of picks, restore autopick list to initial value
      - On itmeout, go through the list and for each `null`, make a random pick
      - Make pick function declarative so that it's easy to understand what's going on
    2. game.js
      - Implement a new pack action called "keep", it does the same thing "pass" does, except it doesn't actually pass the pack.
      - Choose the correct static handlers (callbacks) for the format based on `mkDraftFns` switch function.
    3. bot.js
      - Make it comply with the new semantics
    4. frontend/* and util.js
      - Make necessary changes for the game to recognize this new mode
 2. Make draft functionality, such as `pick` function family polymorphic.
    This turned out to not be needed for this PR (even though I was using this actively during development[1]), but it's something that was considered in a sister-PR[2] and I preserved in this PR.
    1. human.js
      - Use static methods per necessity (so that Human instances can be parametrized by Human functions from Game class at construction time)
    2. game.js
      - Add mk_draft_fns function that can later be used in order to have more esoteric pick functions in a decoupled way (for instance, we can implement Grid draft like that by just implementing JSX and two functions in human.js!)

Automated testing was performed and no regressions were detected.

No new tests were written since this PR only slightly extends the old functionality and no regressions is a good indicator of correctness in this case.

Thorough manual testing was performed.

To understand the process of development better, one can consult development branch, which shall NOT be deleted, because relics of programming should be preserved for insights into design decisions and approaches that didn't work.[1]

P.S.
Sorry for including package bumps and the change to the script, my squash-fu was sup-par.

---
[1]: https://github.com/serokellcao/dr4ft/commits/glimpse-draft-relic
[2]: https://github.com/dr4fters/dr4ft/pull/990#discussion_r415786771

### CODE REVIEW

File: backend/human.js [hunk 1]
```diff
@@ -110,35 +121,105 @@ module.exports = class extends Player {
...
+          return pass_the_pack();
+        else
+          return keep_the_pack();
+      } else if (indexes.length === 1) {
+        burn_a_card(index);
+        return pass_the_pack();
+      } else {
+        burn_a_card(index);
+        if (pack.length === 0)
+          return pass_the_pack();
```
Comment (serokellcao) [inline backend/human.js:208]: Gah, this should say `end_the_round()` for readability, but they are aliases anyway. Sorry for inconsistency, but I'm too lazy to figure out how to amend pull requests without creating new commits. I hope that's not a big deal.

**Comment by HerveH44** [comment]
Hey, to use your word "bikeshedding": I haven't looked much to your PR but in order to keep a bit of continuity in the project, could you use camelCase names for functions, properties, etc? The snake case is looking really different to the rest of the project.

File: backend/human.js (hunk 1 above)
Comment (tooomm) [reply backend/human.js:208]: A new commit is just fine!
GitHub allows to squash them back to one on merge.

**Comment by serokellcao** [comment]
> Hey, to use your word "bikeshedding": I haven't looked much to your PR but in order to keep a bit of continuity in the project, could you use camelCase names for functions, properties, etc? The snake case is looking really different to the rest of the project.

Sorry, will do in a second.

**Comment by serokellcao** [comment]
Please review (and ask me any questions I can help you with), but don't merge at the moment, I need to make commit message (even) prettier.

**Comment by vidbina** [comment]
> Hey, to use your word "bikeshedding": I haven't looked much to your PR but in order to keep a bit of continuity in the project, could you use camelCase names for functions, properties, etc? The snake case is looking really different to the rest of the project.

From the [docs](https://eslint.org/docs/rules/camelcase), I would reckon that ESlint enforces camelCasing by default. Would it be an idea to configure the Travis-CI pipeline to fail on ESlint issues just to make such style-related preferences explicit to contributors?

**Comment by tooomm** [comment]
Hmm, no heroku environment deployed here? :/

**Comment by serokellcao** [comment]
@tooomm, I deployed to http://thoughtflare.memorici.de:22019

**Comment by tooomm** [comment]
The pick counter keeps counting up, even for those picks which you burn.
So you quickly end up with with something like confusing like `Pack 1 - Pick 19/15` displayed and only 6 cards being in your main deck.

In general I think we should help the user whether they are currently making a pick for their deck or are burning cards from the pack and display that somewhere.

**Comment by serokellcao** [comment]
But `Pick 19/15` should never happen!
Can you reproduce?

**Comment by serokellcao** [comment]
Damn me, I just didn't update the deployment with the latest PR, @tooomm, updating it now!
There should never be pick 19/15!

**Comment by serokellcao** [comment]
Okay, now this exact code is tunning at http://thoughtflare.memorici.de:22019.

@tooomm please advice me about what's the best design for discriminating between picks and burns? I'm thinking just green vs red color of the pick number and a text explanation for colorblind compatibility.

**Comment by tooomm** [comment]
Yes, it stops now correctly with "Pick 15/15" and starts over! :)

Two other things:
- The pre-selected default amount of packs should be 9 for a good experience for glimpse
- There is some issue with the pick counter. When doing a glimpse draft with bots and the counter runs down to 0 (doesn't matter if a card is selected for autopick or not), the counter often freezes. However, it still seems to keep on counting as picks are still automatically made and the counter often starts working again with the next automatically made pick.

Hmm, the green/red pick count is a good start. But colour-blindness is a thing to keep in mind, good point! 
There could be e.g. a (coloured?) icon/glyph (with text? or alt text on hover?) next to the pack/pick count. Like a bag, dart, dartboard/bullseye or hair cross to symbolise **keep/take** and a flame for **burn** for example.

**Comment by serokellcao** [comment]
> There is some issue with the pick counter. When doing a glimpse draft with bots and the counter runs down to 0

Wow, well noticed! Will look into it :C

Do we have functional tests?

File: scripts/download_allsets.js [hunk 2]
```diff
@@ -67,22 +67,18 @@ const fetchZip = () => (
...
-      updateDatabase();
-      logger.info("Update DB finished");
-      fs.writeFileSync(setsVersion, version);
-      refreshVersion();
-      await downloadBoosterRules();
-    } else {
-      logger.info("AllSets.json is up to date");
-    }
-  } catch(err) {
-    logger.error(`Couldn't complete AllSets.json check : ${err}`);
```
Comment (mixmix) [inline scripts/download_allsets.js:85]: This change seems unrelated to this feature?

(let me know if new people reviewing parts of a PR is not helpful)

**Comment by serokellcao** [comment]
The more reviews we get the better.
As I said, I sadly squashed changes I needed for development, including
this one and package bumps.

I'm sorry for that, but cherry picking only the implementation commits was
too much of a hustle for me.

:(

**Comment by ZeldaZach** [comment]
FYI you have no reason to squash on your end as I can squash before a merge on GitHub. 

How is this PR coming along?

**Comment by ZeldaZach** [comment]
If you could check on how #990 was handled and follow a similar formatting, would be appreciated. Trying to stay consistent with our approach on adding new components

**Comment by HerveH44** [comment]
Hi  @serokellcao are you waiting for something from me on this PR or do you need any help?

**Comment by serokellcao** [comment]
No, I just need to find the time to update it according to #990.

File: scripts/download_allsets.js (hunk 2 above)
Comment (serokellcao) [reply scripts/download_allsets.js:85]: It is very helpful, also well-spotted!

This change is taken from a patch that I had to merge into my feature branch during development and it got squashed in.

It's already upstream, so it won't cause any trouble.

**Comment by ZeldaZach** [comment]
Ping @serokellcao seeing if you have had any time to possibly look into bringing this up to date. Would love to get it merged!

**Comment by LotusWalker** [comment]
Any update on the status of this implementation? Glimpse drafts are the best!

**Comment by LotusWalker** [comment]
@serokellcao Any update on the status of this implementation? Glimpse drafts are the best!

**Comment by LotusWalker** [comment]
Hi @serokellcao @HerveH44 @vidbina @tooomm @ZeldaZach is there any chance that a Patreon/Paypal donation/contribution could help get this implementation bumped up in priority? Glimpse drafts are the nuts, and Dr4ft would be the only place online to conduct them. Really looking forward to this capability! Thx for all ur efforts.

**Comment by HerveH44** [comment]
Hi @LotusWalker , I'm not part of the project anymore. To answer your question, I don't know what could make this to be merged. It looks pretty close to be ready but as the time passed there would be some conflicts to resolve before merging. Check with @ZeldaZach who has more input on the project.

**Comment by HerveH44** [comment]
Hey, I'll give it a try, using maybe some ideas from @serokellcao PR. I'm considering a simple approach for the UI with a click on a card giving access to 2 buttons, one to pick and one to burn. It seems to me the simplest experience possible to draft it, instead of drag'n'drop for example. If anybody has knowledge/confidence in UI they could help me doing it. I'm starting an ugly PoC but could need a hand on that. the backend is my domain and should be simple enough to implement.

**Comment by mixmix** [comment]
I'm happy to contribute to ui.
Perhaps a couple of SVG buttons

One thing I was wondering is how auto pick will work with this.
Eg ok you select some cards. But then how do you confirm?

If you select 3 cards to burn which of the 3 falls out of selection

**Comment by tooomm** [comment]
As this seems to be stale and the initial author is not repsonding anymore, #1305 probably superses this by now.

**Pull request closed without merging**